search: Redesign the search bar.

evykassirer opened the pull request:
Fixes: https://github.com/zulip/zulip/issues/21798

<details>
<summary>screen captures (outdated)</summary>

updated screenshot with correct option hover color:

<img width="943" alt="image" src="https://github.com/zulip/zulip/assets/5634097/351d7157-e03f-4888-a02a-6a302fa3ca63">

<img width="941" alt="image" src="https://github.com/zulip/zulip/assets/5634097/ece700bc-6224-46f1-bceb-fb199757a80b">



https://github.com/zulip/zulip/assets/5634097/99bb9d46-f1e1-427c-bea0-dc2a0b5dd104

https://github.com/zulip/zulip/assets/5634097/9ff75dea-bd74-426b-8e04-996cd5179b83

https://github.com/zulip/zulip/assets/5634097/16284475-5f12-4edb-875a-ee4e473a99fa

https://github.com/zulip/zulip/assets/5634097/f781bd9b-b3b1-400a-9ea4-b150349dcfc3

https://github.com/zulip/zulip/assets/5634097/34cd9d25-e581-4edc-a1a9-3fb4abb01f41

https://github.com/zulip/zulip/assets/5634097/5ba289b5-37e2-4a08-9be1-3c9fc9f7fde2



</details>

alya commented:
Playing around with this:

1. We were going to have something that looks like a search box in the closed state, right. Having no boundaries around "Search" feels confusing to me.
2. This comment from the other PR still applies:
> Changing the width of the window opens the typeahead.
3. As I resize the window, the typeahead width doesn't change at the same speed as the top part of the box, so it looks glitchy.
![2023-02-09 14 26 42](https://user-images.githubusercontent.com/2090066/217953516-d614e624-4df3-40ab-80a3-811b9b521f55.gif)
4. This comment still applies too; I'm seeing it after I press "Enter" to execute a search:
> there's a gray line that should be behind it, but instead cuts across it.
5. I'm not seeing a hover behavior on the search box in dark theme.

![Screen Shot 2023-02-09 at 2 27 44 PM](https://user-images.githubusercontent.com/2090066/217953656-95162d27-5ef1-4fed-bad4-99481e00cc84.png)
6. I see text spilling out of the typeahead:
![Screen Shot 2023-02-09 at 2 32 46 PM](https://user-images.githubusercontent.com/2090066/217954367-0484f685-f4ac-4b4f-85b9-4ae0dc35ac3c.png)
7. Names spill out of their pills in a narrow window; in `main`, I think they just get covered up by the edge of the window, which is differently bad...

![Screen Shot 2023-02-09 at 2 32 06 PM](https://user-images.githubusercontent.com/2090066/217954741-e2c61417-af33-48fe-bba9-142318d7ed47.png)

That's all I've got at the moment!

evykassirer commented:
> We were going to have something that looks like a search box in the closed state, right. Having no boundaries around "Search" feels confusing to me.

Once we change the background color to grey instead of white, there'll be a boundary. #23724

evykassirer commented:
> 2. Changing the width of the window opens the typeahead.

fixed this

> 3. As I resize the window, the typeahead width doesn't change at the same speed as the top part of the box, so it looks glitchy.

fixed this (same root issue as 2! historic typeahead stuff that doesn't work for this use case)

(working on the others)

evykassirer commented on static/js/message_view_header.js:
```diff
@@ -93,47 +92,16 @@ function append_and_display_title_area(message_view_header_data) {
     }
 }
 
-function bind_title_area_handlers() {
```
these are the old event handlers for search when there was an icon and search bar, new event handlers are in `search.js`

evykassirer commented on static/js/message_view_header.js:
```diff
@@ -146,13 +114,15 @@ function build_message_view_header(filter) {
+        // saves the user a keystroke for quick searches
+        search_string = search_string + " ";
     }
+    $("#search_query").val(search_string);
```
we now always want to reset, even with the empty string, since the searchbar is always visible (sometimes collapsed and sometimes expanded)

evykassirer commented:
4 and 5 -- I'm waiting on Vlad to respond on CZO

>  6. I see text spilling out of the typeahead

I'm no longer seeing this, so I think things I've changed since have fixed it

>  7. Names spill out of their pills in a narrow window; in main, I think they just get covered up by the edge of the window, which is differently bad...

I'm not super familiar with what we're doing with pills -- do you have a sense of what it should look like in this case?

alya commented:
> I'm not super familiar with what we're doing with pills -- do you have a sense of what it should look like in this case?

I don't know if there are places where we solve this well; I see some where we fail to:

![Screen Shot 2023-02-22 at 10 35 23 PM](https://user-images.githubusercontent.com/2090066/220835746-74fc560a-bed4-4c06-a71c-feebffd834ea.png)
![Screen Shot 2023-02-22 at 10 34 36 PM](https://user-images.githubusercontent.com/2090066/220835747-0fe6fdef-b9eb-4ef4-a88d-6e952df43bb3.png)

Maybe abbreviating with a `...` would be best?

evykassirer commented:
Are pills actively being used outside of CZO? Are there engineers/Vlad working on it currently? I'm wondering if labeling the glitch out of scope for this PR would be reasonable.

evykassirer commented:
I've now fixed the border-radius (when search bar is open after a search is complete) and box-shadow issues (simplified it so there's no overlapping shadows) and removed space between search bar and typeahead:

<img width="943" alt="image" src="https://user-images.githubusercontent.com/5634097/221053163-1f2d5d3b-873d-407f-af1d-b0d9371aec75.png">
<img width="948" alt="image" src="https://user-images.githubusercontent.com/5634097/221053199-a4477f78-4108-4977-a234-ddd009ade631.png">

evykassirer commented:
Dark mode now has [correct box-shadow](https://www.figma.com/file/jbNOiBWvbtLuHaiTj4CW0G/Zulip-Web-App?node- id=567%3A51058&t=nulVmvVBfaFHjPB5-0)

![Kapture 2023-02-23 at 15 38 55](https://user-images.githubusercontent.com/5634097/221055955-3630b6f9-3bb6-4b55-b012-a70089dffb30.gif)

evykassirer commented:
remaining known issues:

* if I move away from the window with the typeahead window open then come back, the search bar collapses but the typeahead stays -- need to look into that
* the border radius under the search bar is only broken in dark mode (which doesn't make sense to me, need to look further into that)

evykassirer commented:
> border radius under the search bar

Fixed! Turns out it was also (more subtly) broken in light mode.

terpimost commented:
The things I see so far:

Cross icon is too big
<img width="309" alt="image" src="https://user-images.githubusercontent.com/1903309/222629801-0ab5ad72-bab8-40ed-98af-2f5e10242180.png">

I think is the different search icon and it is slightly misaligned
<img width="205" alt="image" src="https://user-images.githubusercontent.com/1903309/222630018-80102d97-f90d-4608-859e-3465d58a4fc4.png">

Search input and search suggestions should be on a single line:
![CleanShot 2023-03-02 at 23 16 56@2x](https://user-images.githubusercontent.com/1903309/222630179-3ca69a05-5f06-47a5-9f00-72ecb4899104.png)

He click on autosuggestions just closes the search box

In the mobile version lets utilize the full width and cover elements under 
<img width="737" alt="image" src="https://user-images.githubusercontent.com/1903309/222630399-0cc10616-c594-441f-a67f-121fbd1e3eef.png">

something like this:
<img width="967" alt="image" src="https://user-images.githubusercontent.com/1903309/222630438-dc157f65-7080-4386-93b9-bd3eadc4436a.png">

In mobile version the list box lost the shadow for some reason.

In the dark mode mobile version search button has some weird background 
<img width="137" alt="image" src="https://user-images.githubusercontent.com/1903309/222630614-563b51ee-a9c1-4d88-aa5b-49952b8c18c6.png">, it should be just an icon.

Don't forget to recolor the cross icon in the dark version 
<img width="66" alt="image" src="https://user-images.githubusercontent.com/1903309/222630699-6d2f490c-4d1d-45da-a104-74bda29747b5.png">

evykassirer commented:
- [ ] cross icon is too big
- [ ] search icon misaligned
- [x] search input and search suggestions should be on a single line
- [ ] mobile search bar full width ("In mobile version the list box lost the shadow for some reason" will be addressed by this, I gave up on fixing this until I knew what I was turning it into)
- [ ] In the dark mode mobile version search button has some weird background
- [ ] Don't forget to recolor the cross icon in the dark version

> He click on autosuggestions just closes the search box

I don't think I have this behaviour -- can you show me a gif of that?

evykassirer commented on web/e2e-tests/message-basics.test.ts:
```diff
@@ -142,7 +142,7 @@ async function search_and_check(
     expected_narrow_title: string,
 ): Promise<void> {
-    await page.click(".search_icon");
+    await page.click(".search_button");
```
I'll separate out a commit for switching the container to a button and naming it such to be less confusing

evykassirer commented on web/src/message_view_header.js:
```diff
@@ -88,50 +87,15 @@ function append_and_display_title_area(message_view_header_data) {
     }
 }
 
-function bind_title_area_handlers() {
```
Old event handlers that controlled CSS by JS, and it's now mostly pure CSS

evykassirer commented on web/src/message_view_header.js:
```diff
@@ -199,9 +174,7 @@ export function open_search_bar_and_close_narrow_description() {
 }
 
 export function close_search_bar_and_open_narrow_description() {
-    const filter = narrow_state.filter();
```
A lot of callsites were doing this filter check anyways, so I consistently had them do it before calling this function so that the function always does something as its name suggests. (I can also make this a prep commit.)

evykassirer commented on web/src/search.js:
```diff
@@ -90,7 +100,7 @@ export function initialize() {
             return suggestions.strings;
         },
-        parentElement: "#searchbox_legacy",
+        parentElement: "#searchbox_form",
```
Being alongside the other form contents helps it be the right width

evykassirer commented on web/src/search.js:
```diff
@@ -165,14 +175,57 @@ export function initialize() {
+    // shortcuts.
+    // TODO(pre-merge): this still doesn't work great for keyboard navigation by tab,
+    // because the typeahead opens on focus but the search bar doesn't get
+    // fully expanded.
```
I'm having some issues where if I  make this a focus handler, then I get a bug when I search for a new stream and click on it -- instead of closing the search bar and showing me the new stream header, it reopens an empty expanded search bar after going to the new narrow. I don't feel great about this solution though, and am seeking others' thoughts.

evykassirer commented on web/src/search.js:
```diff
@@ -165,14 +175,57 @@ export function initialize() {
+            return;
+        }
+
+        if (page_params.search_pills_enabled) {
```
TODO: I haven't tested anything to do with pills or the new search, and also don't actually know that much about the development plans around it or who to talk to.

evykassirer replied:
#25669 removes that code. It would be great to merge that before this PR gets a thorough review

evykassirer commented on web/styles/dark_theme.css:
```diff
@@ -520,19 +520,75 @@
-        color: hsl(0deg 0% 100%);
-        border: 1px solid hsl(176deg 78% 28% / 60%);
-        background-color: hsl(176deg 49% 42% / 40%);
+    #searchbox,
```
TODO: move these colors into variables

karlstolley commented on web/src/message_view_header.js:
```diff
@@ -162,6 +135,8 @@ export function exit_search() {
         // for "searching narrows", we redirect
         window.location.href = filter.generate_redirect_url();
     }
+    // TODO(pre-merge): why does this not move the cursor out of the search box
```
That might be a delay issue, especially after setting `window.location.href` above. You might experiment with a `setTimeout()` around the focus trigger.

evykassirer commented:
latest push has some pixel perfecting from the Figma mocks, though I also asked several questions [here](https://chat.zulip.org/#narrow/stream/101-design/topic/search.20redesign/near/1575282)

evykassirer commented on web/third/bootstrap-typeahead/typeahead.js:
```diff
@@ -104,7 +104,14 @@
  *   the manual positioning in the show() function.
- *
+ * 
+ * 11. Add `openInputFieldOnKeyUp` option:
```
Not totally sure if this is the best way to do this -- open to ideas!

evykassirer commented on web/styles/dark_theme.css:
```diff
@@ -508,19 +508,74 @@
-        background-color: hsl(0deg 0% 0% / 25%);
-        font-weight: 600;
-    }
+    #searchbox {
```
I might be able to put some of these things into color variables

terpimost commented:
1. Not critical. The search icon doesn't look like it is in a center. 
<img width="712" alt="image" src="https://github.com/zulip/zulip/assets/1903309/65adab24-7cec-4dc2-aabd-1757279450df">
I thought of this icon square as 28x28 square with 16x16 icon in the center. There might be some problem with an icon I prepared or it is just rendered this way. Please check.
Another reason could be that the width of this icon square is 31 instead of 28 
<img width="257" alt="image" src="https://github.com/zulip/zulip/assets/1903309/0b92eb3f-1e74-4cd9-9ba6-ad2c10c47037">


2. Dark mode hover state has white text and icon, which wasn't a design suggestion: 
<img width="215" alt="image" src="https://github.com/zulip/zulip/assets/1903309/d6946855-bab7-4585-8b59-386b08d91e33">
But I understand the potential value of it, because just having hover shadow might not be enough in dark mode. I suggest to have icon 0.75 transparency on hover instead of 100. I don't suggest to change color/transparency of the prompt prompt on hover


3. Why would we need a search icon be a button of itself? 
<img width="429" alt="image" src="https://github.com/zulip/zulip/assets/1903309/62862abe-2c64-436d-8591-3ef8d4a5b46b">


4. Focused state shadow in dark mode is `box-shadow: 0 10px 20px hsla(0, 0%, 0%, 0.25), 0 7px 5px hsla(0, 0%, 0%, 0.1), 0 -7px 9px hsla(0, 0%, 0%, 0.1);` but it should be `box-shadow: 0px 4px 20px rgba(0, 0, 0, 0.3), 0px 1px 5px rgba(0, 0, 0, 0.3);`. In the light theme the focused shadow should be corrected to be `box-shadow: 0px 4px 20px rgba(0, 0, 0, 0.25), 0px 1px 5px rgba(0, 0, 0, 0.1);` instead of `box-shadow: 0 10px 20px hsla(0, 0%, 0%, 0.25), 0 7px 5px hsla(0, 0%, 0%, 0.1), 0 -7px 9px hsla(0, 0%, 0%, 0.1);`


5. Active state placeholder color shouldn't be so bright. Sorry for the confusion. I meant such color for actualy inputed text, while placeholder should be the same as usual: 
<img width="281" alt="image" src="https://github.com/zulip/zulip/assets/1903309/c4e9a50c-de89-4c5d-bba1-d62e211905a3">
In the dark mode it's ok to have search icon be a bit brighter (same as in hover state).
[Updated Figma component](https://www.figma.com/file/jbNOiBWvbtLuHaiTj4CW0G/Zulip-Web-App?type=design&node-id=1716%3A130593&t=3eymGkt12HU5qy7r-1)
<img width="1249" alt="image" src="https://github.com/zulip/zulip/assets/1903309/d9286398-e144-454b-9b19-d0d43aae2828">


6. If I right click on opened suggestions, search bars will be in a weird state: 
<img width="481" alt="image" src="https://github.com/zulip/zulip/assets/1903309/1eef4c6e-8710-41f2-b2bf-2027d7869e73">
And between expended and collapsed state it jiggers through such "half width" state which is not pleasant to the eye:

![CleanShot 2023-06-04 at 15 08 28](https://github.com/zulip/zulip/assets/1903309/61cf635f-f805-4b08-8bc8-b0749d72d85f)
I suggest right now we don't polish the expanding/collapsing animation but make it clean 2 states without anything in between.


7. Close button is to narrow and doesn't touch the right side of the search input: 
<img width="358" alt="image" src="https://github.com/zulip/zulip/assets/1903309/17d81dc1-ca96-477f-b894-47013f95fd79">
It should be 28x28 square and right aligned with right side of input rectangle: 
<img width="196" alt="image" src="https://github.com/zulip/zulip/assets/1903309/bf9b0014-8aeb-4e50-8f03-7adff0d9665a">


8. In Figma I suggested to have a lighter suggestions `color: #595959;` / `hsla(0, 0%, 35%, 1)`, while selected `color` would be black and highlighted match would be black too. Right now I see that typeahead colors in search are reset to `color: inherit;` which isn't great. Here is how it could be: 
<img width="912" alt="image" src="https://github.com/zulip/zulip/assets/1903309/a6d01e66-61de-4668-8848-d78480f62af5">
I updated dark theme design to reflect that idea (suggestions are 75% transparent white, while actual input text and matching list item and matching text fragment color is 100% non transparent white):
<img width="634" alt="image" src="https://github.com/zulip/zulip/assets/1903309/578c44ed-35bc-43c3-8373-9c4462cd9f2e">
Prefix "Search for" should be styled as non matching part while the text from user's input should be highlighted as matching part: 
<img width="274" alt="image" src="https://github.com/zulip/zulip/assets/1903309/e165b847-f4f2-4a76-bfd9-3a96a55e8fae">


9. I couldn't review NO search result page or regular search match result, since it's impossible to get it (jumps to recent conversations):
![CleanShot 2023-06-04 at 15 35 54](https://github.com/zulip/zulip/assets/1903309/41441e63-196d-43b4-9ee8-57529905c93b)

evykassirer commented:
👀 1. I can investigate the icon alignment

✅  2. Updated to 0.75 opacity on hover for the icon and removed the text color change.

![Kapture 2023-06-04 at 12 51 15](https://github.com/zulip/zulip/assets/5634097/0742068d-00d5-4a30-911b-c2745e0e2189)

❓ 3. "Why would we need a search icon be a button of itself?" -- this is for semantic HTML, so that things like screen readers know it's a button that does something. We usually put icons inside containers like this. Do you have any concerns about it?

✅  4. I changed the box shadow to `box-shadow: 0px 4px 20px hsl(0deg 0% 0% / 30%), 0px 1px 5px hsl(0deg 0% 0% / 30%);` and  `box-shadow: 0px 4px 20px hsl(0deg 0% 0% / 25%), 0px 1px 5px hsl(0deg 0% 0% / 10%);` -- it would be really helpful if you could always share colors in hsl, since that's what we do in the CSS and there isn't always a clean conversion between color formats.
 
<img width="206" alt="image" src="https://github.com/zulip/zulip/assets/5634097/241b8060-7c6c-4796-b471-bc8910be00fe">

<img width="202" alt="image" src="https://github.com/zulip/zulip/assets/5634097/4d14b09a-562e-45e6-93b8-fd79a7df2f8f">

✅  5. Removed custom color for placeholder text when the search bar is open in dark mode

<img width="690" alt="image" src="https://github.com/zulip/zulip/assets/5634097/f9a4979b-b94f-4d4e-beaf-484aae282cf7">

🤔  6. These are both things I tried to improve, but essentially it's a problem of us using bootstrap code for the typeahead. To fix this we'd need to do a bunch of dropdown refactoring. I'll ask in design if it's a blocker to shipping the search bar.

✅  7. Updated to 28px by 28px and top/right 0

<img width="128" alt="image" src="https://github.com/zulip/zulip/assets/5634097/3c802b7d-c428-4004-9b17-a73dc3cfc8c1">

✅  8a. "have a lighter suggestions color: #595959; / hsla(0, 0%, 35%, 1), while selected color would be black and highlighted match would be black too."

Updated in light mode and dark mode

<img width="688" alt="image" src="https://github.com/zulip/zulip/assets/5634097/e57332f7-79cb-4217-8c5b-9eb14f56ccf7">

<img width="674" alt="image" src="https://github.com/zulip/zulip/assets/5634097/c993676e-710d-4c3c-9cf8-d046cb2855ab">

🤔 8b. Bolding the part of the search item that matches the input text is going to be a lot more involved and also involve changing how typeahead works. I'll look into how hard that will be.

👀  9. "I couldn't review NO search result page or regular search match result, since it's impossible to get it (jumps to recent conversations)"

Oh dang, thanks for catching this bug! I will start by looking into this.

terpimost commented:
>"Why would we need a search icon be a button of itself?" -- this is for semantic HTML, so that things like screen readers know it's a button that does something. We usually put icons inside containers like this. Do you have any concerns about it?

Because icon is for people who can see and icon isn't a button. The whole input is a button. But icon is just a little image in this case. So I don't think it is semantically correct to put icon as a separate button inside another button. 

Lets take GitHub as example: 
<img width="1024" alt="image" src="https://github.com/zulip/zulip/assets/1903309/c84746b5-1d0c-44c2-b60e-128caba44e10">
Icon is an svg/font/img and not a button

Also it is not a big problem to make something which isn't a button to act like a button, while default browser styles for a <button> might be tricky to cancel. I'm not saying we shouldn't use button just suggesting to use it for buttons with a label.

GitHub toolbar in the comments doesn't use <button> tag... they use custom html tags:
<img width="924" alt="image" src="https://github.com/zulip/zulip/assets/1903309/0c67cabb-9938-47bc-976f-5fcf389a4e53">


> 6. These are both things I tried to improve, but essentially it's a problem of us using bootstrap code for the typeahead.
I guess we could have another pass once we going to deal with compose box and this time that could be postponed, but lets create a follow up issue in this case

> 8b. Bolding the part of the search item that matches the input text is going to be a lot more involved and also involve changing how typeahead works. I'll look into how hard that will be.
Yes, please, take a look. It is not super critical but since we are doing design pass on this element it would be great to make it work and look nice.

evykassirer commented:
I'm pushing a commit that addresses both the issues in `6` which ended up easier than it felt last time I tried.

I've asked on CZO about tips for implementing `8` 

Still investigating `9`, haven't started on `1` yet.

evykassirer commented:
Just pushed a commit that fixes `9`.

I believe the only remaining items are `3` and `8` :)

evykassirer commented:
Addressed `3` by both not renaming the icon container as a button, and also removing that container altogether.

evykassirer commented:
Just pushed a commit that fixes 8 (bold text for the query), though I might end up doing a different technical implementation.

The only remaining comment from Vlad to address is investigating icon centering, which I'm planning to pair on with Karl in the near future.

I'm not currently aware of any other work that needs to be done on this PR.

andersk reviewed (dismissed):
Your escaping changes have introduced cross-site scripting vulnerabilities.

Specifically, you’ve added a lot of paths where a plain text variable is inserted into an HTML string, including paths whose meaning has been changed implicitly without appearing in your diff. **Every time** you insert plain text into HTML, it needs to be escaped for HTML. Any exception is a bug, and any of these bugs are vulnerabilities or could easily become vulnerabilities under seemingly innocuous changes in the future.

Doing that escaping manually is a tedious, error-prone process, which is why we use Handlebars templates instead of doing it manually.

evykassirer commented:
I removed the commit Anders is worried about and followed up on CZO about finding a better solution.

Icon centering was confirmed and looks okay. All of Vlad's comments are now addressed.

I pushed some nice CSS cleanup (removing some legacy absolute positioning) in these commits that I worked on with Karl today.

evykassirer commented:
1. I have a bug with a shadow on hover in a dark theme

I'll look into that, thanks!

2. the whole bar is of a bit different color

To confirm -- @terpimost do you want  the full header to be a different color? I believe the issue is that some of the header isn't the header color and not that some of the header is a different color than the body, but let me know if you want the header to actually be the same color. Someone noticed this on main and we're working on fixing it -- discussion [here](https://chat.zulip.org/#narrow/stream/9-issues/topic/top.20navbar.20background.20dark.20mode/near/1598791).

3. No search results could be improved

Great callout! I didn't notice that part of the design. I can work on that as another commit in this PR, or a separate PR.

4. We lost space between user pill and the text

I will look into that.

5. Close button takes full width but not height, which is wrong

I will look into that

6. The suggestion about bold was not only to change font-weight but also contrast of a color, right now it is jut font-weight

I talked about this a bit on [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/search.20redesign/near/1598548) but I haven't changed anything about the bold or font weight stuff yet. Are you seeing different behaviour than on CZO? Or maybe you're talking about something different than that?

7. Screenshot of previous request about colors

Are you not seeing the color changes? I see them on my version locally. Or am I misinterpreting your comment?

Thank you very much for taking a look! I'm excited to be getting close  :)

terpimost commented:
>  I believe the issue is that some of the header isn't the header color 

Correct, I was talking about the same issue.

> haven't changed anything about the bold or font weight stuff yet

There are 2 issues here. One is what to bold... which is might be hard for the very first search result as I understand. Second is how to bold - I suggest to not only bold but change a color of that so bolded text would be darker in light mode and lighter in dark mode: 
see here **de** is of a black color, while the text of suggestions is dark grey
<img width="141" alt="image" src="https://github.com/zulip/zulip/assets/1903309/c4bccc44-e2c7-451b-908a-b8463b13981c">

evykassirer commented:
@terpimost the close button looks to me like it takes full height, am I missing something?

<img width="132" alt="image" src="https://github.com/zulip/zulip/assets/5634097/6b2e26d3-8ffa-46b2-b138-439e1ddcfe28">

evykassirer commented:
> Second is how to bold - I suggest to not only bold but change a color of that so bolded text would be darker in light mode and lighter in dark mode:

Ah I see! I haven't changed anything relating to the bolding, but it sounds like you want to change the _existing_ bolding to be more contrasting _as well as_ having the top result include the bold text? I can remember that when I get working on that.

evykassirer commented:
fixed:

* 1 has been fixed in this PR
* 2 has been fixed in a separate PR and merged into main
* 4 has been fixed in this PR

3 is waiting on @terpimost's  design, 5 is waiting on Vlad's response to my question above

and 6/7 I'll work on in a separate PR for now, to set up the bolding.

terpimost commented:
>  want to change the existing bolding to be more contrasting as well as having the top result include the bold text?

Yes, top search result isn't that critical, while more contrasting current bolding is very helpful

> 3 is waiting on @terpimost's design

Discussing in Zulip chat

karlstolley commented on web/styles/compose.css:
```diff
@@ -892,6 +892,7 @@ a.compose_control_button.hide {
 
     &.typeahead {
         background: hsl(0deg 0% 100%);
+        right: 0;
```
I think this duplicates what's already set in `search.css`

evykassirer replied:
yup! thanks, removed

karlstolley commented on web/styles/dark_theme.css:
```diff
@@ -521,19 +521,31 @@
+          and dark theme shows it mostly through changing color
+          of the search button and placeholder text. */
+        .navbar-search:not(.expanded) #search_arrows:hover {
+            /* Light theme has two different (very similar) colors in the box
```
It would be great (in another PR) to do a roundup of similar colors like this and run them by @terpimost to see where we could consolidate and make things uniform.

evykassirer replied:
[asked here](https://chat.zulip.org/#narrow/stream/101-design/topic/search.20redesign/near/1602346)

karlstolley commented on web/styles/dark_theme.css:
```diff
@@ -668,6 +680,10 @@
     }
 
+    #searchbox_form .dropdown-menu ul {
+        background-color: inherit;
```
One thing to watch for is when it's better to genuinely inherit a background color (which gives the element a background color for real) and when it's better to just set it to [`unset`](https://developer.mozilla.org/en-US/docs/Web/CSS/unset)--effectively allowing the background of the containing element to just show through). I don't know what's the better option here, but if `unset` works, it's probably better to stick with it.

evykassirer replied:
Cool, new CSS knowledge for me! `unset` does work :)

karlstolley commented on web/styles/search.css:
```diff
@@ -2,144 +2,219 @@
+    .navbar-search {
+        margin: 5.5px 0;
+        border-radius: 5px;
+        position: absolute;
```
It's a little hard to mess with this in the Firefox inspector, but elsewhere in this file, `.navbar-search` is floated--which won't have any effect if it's actually an absolutely-positioned element.

evykassirer replied:
Yup! Good catch. Removed the float.

karlstolley commented on web/styles/search.css:
```diff
@@ -2,144 +2,219 @@
+
+    .search_close_button {
+        width: 28px;
+        background: none;
```
There's a lot of reset-looking stuff here. Again, the browser inspector isn't playing nice for me to look at this, but it might be worth seeing if any of these can be safely removed.

evykassirer replied:
Ah yeah this is blindly removing default button styles, but they're not all needed here. I just removed about half of them!

karlstolley commented:
All right, Evy. I've reviewed the CSS here. There are a few small comments and things you might want to check for. 

It's really tricky trying to poke at this in the browser, because (as I'm guessing you know) even manually adding classes like `.expanded` or inline `display: block` styles to hold things open quickly collapse thanks to our solid but in this case a bit annoying event-handling to close things up and off.

That said, when you next want me to take a look at this, please let me know.

evykassirer commented:
@terpimost can you respond to [this comment](https://github.com/zulip/zulip/pull/24345#issuecomment-1610002204) above? I think I've otherwise addressed your comments in this PR or by opening other PRs.

terpimost commented:
There should be no focus outline on click and it should be of the full height anyway :)
<img width="124" alt="image" src="https://github.com/zulip/zulip/assets/1903309/c8b81550-0d09-433c-9123-e90e547b22aa">

Please remove custom focus/active class and replace it with `focus-visible` - works only when focused via a keyboard.
Please make the height of the button to utilize the full height of the input field

In the dark theme I do see some strange visual bug:
<img width="149" alt="image" src="https://github.com/zulip/zulip/assets/1903309/28d90f24-695a-4d3e-80b0-12f1ea610f0e">

Input text is black in the dark theme for some reason: 
<img width="237" alt="image" src="https://github.com/zulip/zulip/assets/1903309/96d6cbf9-23e0-4a64-af40-2df3d1ce0683">

evykassirer commented:
Sigh, that dark theme bug came from #26131 which was merged two weeks ago. This is an unfortunate effect of leaving a PR open this long while other work is being done. Thanks for catching that!

I updated the focus thing as well:

```diff

diff --git a/web/styles/dark_theme.css b/web/styles/dark_theme.css
index 4e0b708a62..6f307e7fb8 100644
--- a/web/styles/dark_theme.css
+++ b/web/styles/dark_theme.css
@@ -482,8 +482,7 @@
     select,
     .pill-container,
     .user-status-content-wrapper,
-    #custom-expiration-time-input,
-    #searchbox #search_query {
+    #custom-expiration-time-input {
         background-color: hsl(0deg 0% 0% / 20%);
         border-color: hsl(0deg 0% 0% / 60%);
         color: inherit;
diff --git a/web/styles/search.css b/web/styles/search.css
index 8ec121aa24..4a12744e0c 100644
--- a/web/styles/search.css
+++ b/web/styles/search.css
@@ -32,10 +32,14 @@
 
     .search_close_button {
         width: 28px;
+        height: 100%;
         background: none;
         border: none;
         opacity: 0.5;
         line-height: 0;
+        &:not(:focus-visible) {
+            outline: none;
+        }
 
         &:hover {
             opacity: 1;
```

terpimost commented:
If I'm focused in the field and hit `Tab` it just closes the input. Should it be so? Right now it acts like `Enter` key

evykassirer commented:
Tab acting like enter key doesn't make that much sense to me, so I prefer the way it works in this PR. Thanks for noticing it changed though -- would you prefer tab acting like enter?

evykassirer commented:
Confirmed on CZO that current behaviour is fine. Vlad would like to have Tab move the focus to the close button, but that doesn't seem blocking.

karlstolley commented:
Okay Evy! I found a couple of things:

* One big regression I note here is that the `/` key doesn't properly refocus the search input box from all search views or certain narrows. For example, searching for just a text string, or for posts from a `sender` renders the `/` key unresponsive. Something about those narrows is off, even if you go directly to the URL, e.g., http://localhost:9991/#narrow/sender/11-Iago That seems important to fix prior to CZO burn-in. Keyboard users, like me, instinctively whack that thing to refine the search terms when the results aren't satisfactory.
* The Tab key does seem to focus the Close button on a second strike. I still don't think that's ideal, but it is frustration-discoverable :) ("[Tab] You should go to close! [Tab] Oh, okay. Cool.")

evykassirer commented:
I wasn't able to replicate the `/` regression for `sender` views, but I have replicated it for text string searches so I'll start there :)

evykassirer commented:
I think this should fix it, do you see the issue still after the most recent push?


```diff

diff --git a/web/src/search.js b/web/src/search.js
index 6adcc1301c..3f6f637694 100644
--- a/web/src/search.js
+++ b/web/src/search.js
@@ -153,10 +153,9 @@ export function initialize({on_narrow_search}) {
 
 export function initiate_search() {
     // Don't reset the search bar if it's already open.
-    if ($(".navbar-search.expanded").length) {
-        return;
+    if (!$(".navbar-search.expanded").length) {
+        message_view_header.open_search_bar_and_close_narrow_description();
     }
-    message_view_header.open_search_bar_and_close_narrow_description();
     open_typeahead();
 }
```

evykassirer commented:
Added a commit to remove the custom handling for Tab in the typeahead, does that seem like the right solution?

evykassirer commented:
@karlstolley ready for your review again :)

karlstolley commented:
Hi @evykassirer! Okay, the `/` issues seem to be taken care of. The Tab-handling seems right functionally. Let me look at the code now and I'll comment if I have anything to suggest.

karlstolley commented:
Hmmmm. I can't seem to replicate it now, but somehow I got into a state on Recent conversations where the usual Tab traps that keep a user tabbing endlessly through the filters or individual conversation rows stopped working, so that it was possible to tab from the filters box into the first message row, and on down through the entire message list. The only real problem with that (other than a behavioral change that should be discussed if it's intentional, but I can't seem to replicate now) is that the tooltips on the Time column didn't clear.

karlstolley commented on web/src/message_view_header.js:
```diff
@@ -87,65 +86,35 @@ function append_and_display_title_area(message_view_header_data) {
+    let search_string = narrow_state.search_string();
+    if (search_string !== "" && !narrow_state.filter().is_search()) {
+        // saves the user a keystroke for quick searches
+        search_string = search_string + " ";
```
This is a really nice touch.

karlstolley commented on web/src/search.js:
```diff
@@ -111,11 +133,30 @@ export function initialize({on_narrow_search}) {
-    message_view_header.open_search_bar_and_close_narrow_description();
-    $("#search_query").typeahead("lookup").trigger("select");
+    // Don't reset the search bar if it's already open.
+    if (!$(".navbar-search.expanded").length) {
```
This is the Boolean example with `.navbar-search.expanded`

evykassirer replied:
Switched to `if ($(".navbar-search.expanded").length > 0)` :)

karlstolley commented on web/styles/search.css:
```diff
@@ -2,150 +2,232 @@
+        margin: 5.5px 0;
+        border-radius: 5px;
+        position: absolute;
+        color: hsl(0deg 0% 0% / 50%);
```
I've seen this color a bunch just in the context from this PR; it sticks out next to a proper CSS variable. Probably not worth clouding the PR here, but putting repeat colors like this behind a variable could be a good follow-up PR.

evykassirer replied:
I can only find it once -- can you point me to where else you were seeing it?

karlstolley replied:
I must've gotten goofed up and ignored that one instance was from a removed line.

karlstolley commented on web/styles/search.css:
```diff
@@ -2,150 +2,232 @@
         @media (width < $md_min) {
-            margin-left: 40px;
+            #searchbox-input-container {
+                width: 40px;
```
It might be good to put the source of dimensions like this in a comment.

evykassirer replied:
I got the dimensions from the Figma file, would you recommend adding that in a comment? That's where I get most of my dimensions.

evykassirer replied:
Added a comment here!

karlstolley commented on web/styles/search.css:
```diff
@@ -2,150 +2,232 @@
-            @media (width < $sm_min) {
-                top: 5px;
+        box-shadow:
+            0 4px 20px hsl(0deg 0% 0% / 25%),
```
Are these colors different from the box-shadows above that have variables assigned to them? I'm guessing so, but there's another case where variables would be an asset on wrangling all of these colors.

evykassirer replied:
Yeah this one is shadow around the whole search situation (input field and typeahead) when it's open, to create some visual separation from the rest of the page.

Can you recommend what to call the colors? The first is an bigger shadow and the second is a tighter one. I made them blue and red below to show what they do:

<img width="850" alt="image" src="https://github.com/zulip/zulip/assets/5634097/8a8d5871-5b0f-4527-9529-b2dcb5a9e1e6">

evykassirer replied:
Updated to

```css

        box-shadow:
            0 4px 20px var(--color-search-shadow-wide),
            0 1px 5px var(--color-search-shadow-tight);

```

karlstolley commented on web/styles/search.css:
```diff
@@ -2,150 +2,232 @@
-        font-weight: 300;
-        line-height: var(--header-height);
+        font-weight: 400;
+        height: 28px;
```
This 28px height shows up a couple of places nearby. Might also be good to put in a variable, like `--header-height` above, which seems to be gone now?

evykassirer replied:
header height is still around!

<img width="355" alt="image" src="https://github.com/zulip/zulip/assets/5634097/ff53295d-041a-4aed-9f89-77bccafc058e">

Added this

```css
    /*
    Height of the search box, which appears in the header.
    On very small screens, the search box's height becomes
    the height of its parent.
    */
    --search-box-height: 28px;
```

karlstolley commented on web/styles/search.css:
```diff
@@ -2,150 +2,232 @@
-            top: 0;
+        #searchbox-input-container,
+        .search-input {
+            height: var(--header-height);
```
Oooh, here's a height with a variable on `.search-input`. Perhaps that value needs to be adjusted within the relevant media queries, if that's changing?

evykassirer replied:
Oooo yeah cool, updated to this:

```css
    /*
    Height of the search box, which appears in the header.
    On very small screens, the search box's height becomes
    the height of its parent (i.e. the header height).
    */
    --search-box-height: 28px;
    @media (width < $sm_min) {
        --search-box-height: var(--header-height);
    }
```

karlstolley commented on web/styles/search.css:
```diff
@@ -2,150 +2,232 @@
+    .typeahead-menu {
         display: flex;
-        align-items: center;
+        flex-direction: column;
```
Is there a reason to flex this with a column direction, and not just go with the normal document flow (`display: block`)? I'm not noticing a difference if I switch off `display: flex` on this in the Inspector, but maybe it's subtle? Or there's some case where we actually need flex properties?

evykassirer replied:
Oh interesting -- so I thought I needed this for the `search: Wrap long search suggestions.` commit, but maybe the flex wrap on `search_list_item` was enough. I also don't see any difference.

karlstolley commented on web/styles/search.css:
```diff
@@ -2,150 +2,232 @@
+        padding: 3px 30px;
+
+        .search_list_item {
+            display: flex;
```
I'm not sure about this flexbox either, since `.search_list_item` appears to have a single child, `span`--maybe that's what ought to be flexed in order to solve the still-present spillover using `overflow-x: hidden`? You might also need a gap to keep "Search for" from colliding with the contents of the `<strong>` element.

Before:
<img width="605" alt="spillover" src="https://github.com/zulip/zulip/assets/170719/547a3528-658a-4c29-9ec5-626d97e71016">

After:
<img width="473" alt="contained" src="https://github.com/zulip/zulip/assets/170719/0b243b10-ffe4-4b91-93a8-b3506f543bd2">


The changes:
<img width="295" alt="flex-changes" src="https://github.com/zulip/zulip/assets/170719/51451a48-b17d-4909-9208-7556895cdea1">

evykassirer replied:
Hm, when I do that things get cut off when I don't want them to (first is your version, second is my original one)

<img width="763" alt="image" src="https://github.com/zulip/zulip/assets/5634097/09d4f379-da8b-4bb4-8184-dd6d81d179ff">

<img width="760" alt="image" src="https://github.com/zulip/zulip/assets/5634097/c7cba385-e977-4feb-965f-f5bf0ff133ba">

But I am seeing that it's still not wrapping great for that kind of input (and another I just got)

<img width="842" alt="image" src="https://github.com/zulip/zulip/assets/5634097/099f869d-b986-458f-8837-4d1cd7e117ba">

Can we pair on this together?

evykassirer replied:
Addressed during pairing!

evykassirer commented on web/src/search.js:
```diff
@@ -72,6 +77,23 @@ export function initialize({on_narrow_search}) {
         // the search bar as soon as the user hits Esc.
         on_escape: message_view_header.exit_search,
+        openInputFieldOnKeyUp() {
+            if ($(".navbar-search.expanded").length === 0) {
```
Thanks for noticing that! Switched the other one to `if ($(".navbar-search.expanded").length > 0)` :)

evykassirer commented:
Changed some things from your comments (noted in relevant threads) and pushing that now. 

A few things I asked to further clarification on.

Thanks so much for your review!

evykassirer commented:
Addressed all your comments @karlstolley !

karlstolley commented on web/third/bootstrap-typeahead/typeahead.js:
```diff
@@ -495,6 +515,13 @@ import {get_string_diff} from "../../src/util";
           }
+          if (this.options.openInputFieldOnKeyUp !== null && !this.shown) {
+            // If the typeahead isn't shown yet, the `lookup` call will open it.
+            // Here we make do a callback to the input field before we open the
```
```suggestion
            // Here we make a callback to the input field before we open the
```

karlstolley commented on web/third/bootstrap-typeahead/typeahead.js:
```diff
@@ -104,7 +104,26 @@
+ * 11. Add `openInputFieldOnKeyUp` option:
+ *  
+ *   If the typeahead isn't shown yet, the `lookup` call in the keyup
+ *   handler will open it. Here we make do a callback to the input field
```
```suggestion
 *   handler will open it. Here we make a callback to the input field
```

karlstolley commented:
Whew! All right, @evykassirer -- I suggested a couple of nits in comment wording, but otherwise I think this is ready for some time on CZO, so I'm marking it as such.

evykassirer commented:
Fixed those two things! And also fixed them in #26139 since that's where those commits are expected to be merged.

timabbott commented:
Alya and I played with this and found a couple bugs/quirks:
* There's no border/difference between the text box section and the options to select section of the search here; somehow that feels a bit off to me but I'm not sure what to propose as a change. 
![image](https://github.com/zulip/zulip/assets/2746074/cae3a961-779e-426a-a806-f43eaba5173c)

* If I got to a group DM thread, and click to open search, and then click on the "X" to close it, there's a weird flicker that appears like the animation/update got split into two steps.
* Should there be a stronger hover behavior or tooltip when hovering over the search box? I feel like a tooltip might be overkill but somehow I feel like just the magnifying glass changing color feels a little weak? Unsure. Oh! In light theme, there's an on-hover shadow (but no on-hover change to the magnifying glass!) and in the dark theme it's the opposite; this is probably not intentional.
* The light them contrast feels like it's in the uncanny valley where it almost looks like it's the same but it slightly different but it's too faint of a contrast to easily make out.
![image](https://github.com/zulip/zulip/assets/2746074/4d16327e-2c0d-49db-b438-6d1ad0ed9094)
* The bottom borders seem broken in some way in narrow windows:
![image](https://github.com/zulip/zulip/assets/2746074/f5a65ce6-991d-49d7-919d-cfef62d9c9c8)

evykassirer commented:
> If I got to a group DM thread, and click to open search, and then click on the "X" to close it, there's a weird flicker that appears like the animation/update got split into two steps.

Ah, this happens any time the X is clicked. There's some weird stuff around the typeahead closing before the search bar closes. I fixed this with:

```diff
diff --git a/web/src/message_view_header.js b/web/src/message_view_header.js
index 2b68bcc54e..e337672791 100644
--- a/web/src/message_view_header.js
+++ b/web/src/message_view_header.js
@@ -163,10 +163,14 @@ export function open_search_bar_and_close_narrow_description() {
 }
 
 export function close_search_bar_and_open_narrow_description() {
-    $("#search_query").val("");
-    $(".navbar-search").removeClass("expanded");
-    $("#message_view_header").removeClass("hidden");
     if ($("#search_query").is(":focus")) {
+        // Blur first so that the typeahead closes at the same time as
+        // the search bar. Blurring will trigger this function to be
+        // called again.
         $("#search_query").trigger("blur");
+        return;
     }
+    $("#search_query").val("");
+    $(".navbar-search").removeClass("expanded");
+    $("#message_view_header").removeClass("hidden");
 }
```

evykassirer commented:
> The light them contrast feels like it's in the uncanny valley where it almost looks like it's the same but it slightly different but it's too faint of a contrast to easily make out.

Darn, I've messed around with this before and thought I fixed it. I'll take a look again.

evykassirer commented:
posted other followups on CZO here: https://chat.zulip.org/#narrow/stream/101-design/topic/search.20redesign

timabbott commented:
Doing a test-deploy of this on chat.zulip.org so we can get some feedback.

terpimost commented:
I see now.  As @evykassirer suggested we can try to solve that with minimum change:
dark mode:
![CleanShot 2023-09-04 at 19 23 34@2x](https://github.com/zulip/zulip/assets/1903309/f67b38f6-fe69-49bb-9ef5-45462d45304c)


```cs
    border-top-width: 1px;
    border-top-color: rgba(0, 0, 0, 0.35);
    border-radius: 0px 0px 4px 4px;
```

light mode: `border-top-color: rgba(0, 0, 0, 0.1);`

![CleanShot 2023-09-04 at 19 27 10@2x](https://github.com/zulip/zulip/assets/1903309/0c3be27b-9cf7-4bc5-abb1-83177678f0ee)


Ideally I would add backdrop to darken all content and indicate that active search input is a modal... but it might be hard at the current step. Do you think we should do that?

Other apps doing different things:
Google chrome does nothing in light mode: 
<img width="1593" alt="image" src="https://github.com/zulip/zulip/assets/1903309/a45a51cf-0189-4440-8db2-6f284c7c7f17">
<img width="1335" alt="image" src="https://github.com/zulip/zulip/assets/1903309/5087888e-1a5f-410c-ac9e-498d99443923">

Yandex does that beautifully (it has nothing in the background like zulip ... no borders under it):
<img width="1024" alt="image" src="https://github.com/zulip/zulip/assets/1903309/f1dcc018-77c5-45d8-a38a-0c70a2ff6c74">

Youtube doesn't look beautiful: 
<img width="792" alt="image" src="https://github.com/zulip/zulip/assets/1903309/284a053e-f1cd-48cd-85a1-7f2747326721">

Duck Duck go:
<img width="742" alt="image" src="https://github.com/zulip/zulip/assets/1903309/643b7b20-0c00-4e07-8ef6-35fc9e59a8af">
<img width="754" alt="image" src="https://github.com/zulip/zulip/assets/1903309/a33e89fe-0380-4d36-a258-a196cd99b791">

evykassirer commented:
Updates in the recent push:

* Rounding the borders of the focus outline for the X button and updates on styling for the outline to not overlap nearby elements
* Updated the tab keypress commit to be custom for search
* Added a new commit `typeahead: Don't act on blurs that change focus within the parentElement.` that stops Tab from closing the searchbox, and some blur/keydown handlers elsewhere to ensure that Tab from the close button does still close the searchbox
* Simplified `close_search_bar_and_open_narrow_description` and moved the logic to avoid the narrowing dropdown visual jump to its callsite -- this code is still in the temp commit at the end of this PR 
* Added the border top to the typeahead

alya commented:
It's great to see this moving along!

Playing with the PR, I noticed that the placeholder word "Search" is in a darker font (in light theme) than actual search terms, which seems weird to me. Is this intentional?

<img width="254" alt="Screen Shot 2023-09-08 at 11 56 59 AM" src="https://github.com/zulip/zulip/assets/2090066/1ae9137c-ba68-41c6-823a-805a5bdfda72">
<img width="307" alt="Screen Shot 2023-09-08 at 11 56 20 AM" src="https://github.com/zulip/zulip/assets/2090066/1d84b4b5-1b50-438f-85f9-493e33ea3842">

alya commented:
Something seems a bit off with the animation when I click `x` to close to search box. I *think* the typeahead is getting shifted to the right, and then immediately closed. It would probably look more smooth if we just closed the typeahead dropdown directly.

alya commented:
I don't know how critical this is, but if I click on the search box and start typing very quickly, the start of what I type disappears when the search box expands. I actually found this easier to replicate by tabbing to the search box, though it's kinda awkward to figure out how to get the keyboard nav to the right place for that.

alya commented:
Mm, I'm seeing problems with overlapping icons in narrow windows:

<img width="719" alt="Screen Shot 2023-09-08 at 12 36 23 PM" src="https://github.com/zulip/zulip/assets/2090066/802352b4-43c3-4792-9d81-3d544ce3dcf7">

alya commented:
> Something seems a bit off with the animation when I click `x` to close to search box. I _think_ the typeahead is getting shifted to the right, and then immediately closed. It would probably look more smooth if we just closed the typeahead dropdown directly.

I see this one was [noted on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/.E2.9C.94.20typeahead.20changes.20for.20search/near/1585014) a while back as well.

evykassirer commented:
> I noticed that the placeholder word "Search" is in a darker font (in light theme) than actual search terms, which seems weird to me. Is this intentional?

That doesn't seem intentional! I updated it to have a similar situation as the dark mode, with the placeholder text being a lighter color than the color for the input and also dropdown options.

> > Something seems a bit off with the animation when I click `x` to close to search box. I _think_ the typeahead is getting shifted to the right, and then immediately closed. It would probably look more smooth if we just closed the typeahead dropdown directly.
> 
> I see this one was [noted on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/.E2.9C.94.20typeahead.20changes.20for.20search/near/1585014) a while back as well.

Yeah this has been a frustrating issue to fix because of how the code is set up here as well as the functionality I want out of closing the typeahead in various circumstances. I feel a bit like I've been doing a bit of wackamole, but hopefully the thing I just changed (still in the temp commit at the end of the PR) works this time!

> I don't know how critical this is, but if I click on the search box and start typing very quickly, the start of what I type disappears when the search box expands. I actually found this easier to replicate by tabbing to the search box, though it's kinda awkward to figure out how to get the keyboard nav to the right place for that.

Ah, great find! I was able to replicate it easily enough with a click. Fixed with this:

```diff 

diff --git a/web/src/message_view_header.js b/web/src/message_view_header.js
index d482cadff1..cb4478a475 100644
--- a/web/src/message_view_header.js
+++ b/web/src/message_view_header.js
 
 export function open_search_bar_and_close_narrow_description() {
-    reset_searchbox_text();
+    // Preserve user input if they've already started typing, but
+    // otherwise fill the input field with the text operators for
+    // the current narrow.
+    if ($("#search_query").val() === "") {
+        reset_searchbox_text();
+    }
     $(".navbar-search").addClass("expanded");
     $("#message_view_header").addClass("hidden");
 }
```

> I'm seeing problems with overlapping icons in narrow windows.

I just introduced that bug when trying to refactor something. Thanks! Fixed it.

evykassirer commented on web/tests/search.test.js:
```diff
@@ -331,12 +331,11 @@ run_test("initiate_search", () => {
     assert.ok(is_searchbox_text_selected);
     // test that we append space for user convenience
-    narrow_state.filter = () => ({is_search: () => false});
-    search.initiate_search();
```
It doesn't make sense to `search.initiate_search();` while it's already open -- to test for the appended space, we need to initiate search from a closed search state, which we were already doing.

So I updated the test to only initiate search once

evykassirer commented:
Ready for re-review :)

alya commented:
The unread count looks odd on really narrow screens; I think we should just hide it in this situation:
<img width="537" alt="Screen Shot 2023-09-12 at 2 06 38 PM" src="https://github.com/zulip/zulip/assets/2090066/b50cde83-45c9-48f8-9a38-eca66dbd5f1b">

alya commented:
Otherwise looks good to me in manual testing. Thanks for all the fixups!

evykassirer commented:
Great catch -- just pushed a fix to that :)

timabbott commented:
Test-deploying on chat.zulip.org. (The test failure was a known flake in `main`).

timabbott commented:
Apparently it didn't actually deploy yesterday, fixing now.

timabbott commented:
@evykassirer can you clean up the last "temp" commit at this point? I think the UI is getting close to completion and I'm not sure how to review it.

evykassirer commented:
squashed the temp commit into the main commit and pushed  :)

timabbott commented on web/third/bootstrap-typeahead/typeahead.js:
```diff
@@ -539,6 +544,11 @@ import {get_string_diff} from "../../src/util";
+      // hide the typeahead.
+      if ($(e.relatedTarget).parents(this.options.parentElement).length > 0) {
+        return;
+      }
```
Can you provide an example of this and an argument for this change won't potentially make weird things happen with other typeahead components like the compose box? I'm a bit confused as to what this does and unlike the other typeahead changes, it's not optional.

evykassirer replied:
I can copy the example from above here, or would you want something more?

> *   This allows us to have things like a close button, and be able
> *   to move focus there without the typeahead closing.

Essentially I want to allow focus to move within the input field without the typeahead closing. I don't think this is relevant to anything else becaues `parentElement` is only used for search.

timabbott replied:
Ohh, OK, it's an option we added that we only pass there.

timabbott commented:
I think this lgtm aside from https://github.com/zulip/zulip/pull/24345#discussion_r1330838138

evykassirer commented:
I found a bug where the typeahead isn't properly narrowed after opening the search bar (it shows the generic options until the user takes some action in the input field.)

timabbott commented:
Merged, thanks @evykassirer and @YashRE42 for all the work on this!
